modules/command-engine/core/src/main/scala/surge/internal/kafka/KafkaProducerActorImpl.scala

```diff
@@ -225,18 +249,44 @@ class KafkaProducerActorImpl(
   private def handleProcessingProducerMessage(message: KafkaProducerActorImpl.KafkaProducerActorMessage, state: KafkaProducerActorState): Unit = {
     message match {
       case msg: Publish =>
-        context.become(processing(state.addPendingWrites(sender(), msg)))
+        val maybeTrackerWithExpiry = publishTrackerStateManager.tracker(msg.batchId, state)
+        maybeTrackerWithExpiry match {
+          case Some(trackerWithExpiry) =>
+            if (!trackerWithExpiry.tracker.dataWasPublished()) {
+              if (!trackerWithExpiry.expired) {
+                // Avoid duplicating a write that is already pending.
+                if (!publishTrackerStateManager.stateHasPendingWrites(msg.batchId, state)) {
+                  context.become(processing(state.addPendingWrites(sender(), msg)))
+                } else {
+                  context.become(processing(state))
+                }
+              } else {
+                context.become(processing(state.stopTracking(msg)))
+                context.self ! EventsFailedToPublish(
+                  reason = new RuntimeException(s"PublishTracker expired, not attempting another publish for ${msg.batchId}"),
+                  originalSenders = Seq(SenderWithTrackingId(sender(), msg.batchId)))
+              }
+            } else {
+              context.become(processing(state))
+              context.self ! EventsPublished(Seq(SenderWithTrackingId(sender(), msg.batchId)))
```

Comment:
We can just directly reply with success in this case rather than sending a message to ourself to reply successful to avoid an extra step/unnecessary message in the mailbox to process

Reply:
done